### PR TITLE
Use door stop command in place of toggle

### PR DESCRIPTION
**Pull request by mariusmuja**
Currently the door stop button actually send a door toggle command, which does stop the door when it's opening, but it reverses the direction when the door closes. 

I found that there is a stop command that actually stops the door in both cases. It's still possible to reverse the direction while closing the door by using the "open" button.

### CODE REVIEW

File: base.yaml [hunk 1]
```diff
@@ -177,7 +177,7 @@ button:
           id($id_prefix).query_openings();
 
   - platform: template
-    id: ${id_prefix}_test
+    id: ${id_prefix}_sync
```
Comment (mariusmuja) [inline base.yaml:183]: I sneaked this in to fix a bug I introduced in the previous PR.

**Comment by bdraco** [comment]
🔥

**Comment by bdraco** [comment]
Will test after breakfast

**Comment by bdraco** [comment]
I got a crash when flipping the light back and forth 

```
[09:01:28][V][json:036]: Attempting to allocate 512 bytes for JSON serialization
[09:01:28][V][json:056]: Size after shrink 76 bytes
[09:01:28][D][ratgdo:217]: [162389] Encode for transmit rolling=000060f fixed=0200000539 data=00000081
[09:01:28][V][ratgdo:252]: Counter: 1551 Send code: [550100443FBFDD6FF7FF7F66397E92DC1F27F9]
[09:01:28][D][number:012]: 'Rolling code counter': Sending state 1552.000000
[09:01:28][V][json:036]: Attempting to allocate 512 bytes for JSON serialization
[09:01:28][V][json:056]: Size after shrink 88 bytes
[09:01:28][V][component:204]: Component light took a long time for an operation (0.06 s).
[09:01:28][V][component:205]: Components should block for at most 20-30ms.
[09:01:28][D][ratgdo:151]: [162452] received mine: rolling=000060f fixed=0200000539 data=00000081
[09:01:28][V][ratgdo:369]: Light state OFF (0)
[09:01:28][V][json:036]: Attempting to allocate 512 bytes for JSON serialization
[09:01:28][V][json:056]: Size after shrink 76 bytes


WARNING smallgarageratgdo.local: Connection error occurred: [Errno 54] Connection reset by peer
INFO Processing unexpected disconnect from ESPHome API for smallgarageratgdo.local
WARNING Disconnected from API
INFO Successfully connected to smallgarageratgdo.local

```

**Comment by bdraco** [comment]
Its reproducible by flipping the light back and forth many times quickly

```
[09:03:21][D][light:036]: 'Light' Setting:
[09:03:21][D][light:047]:   State: OFF
[09:03:21][V][json:036]: Attempting to allocate 512 bytes for JSON serialization
[09:03:21][V][json:056]: Size after shrink 76 bytes
[09:03:21][D][ratgdo:217]: [104537] Encode for transmit rolling=0000657 fixed=0200000539 data=00000081
[09:03:21][V][ratgdo:252]: Counter: 1623 Send code: [5501000826934534924D24842BFD97F81F6F7B]
[09:03:21][D][number:012]: 'Rolling code counter': Sending state 1624.000000
[09:03:21][V][json:036]: Attempting to allocate 512 bytes for JSON serialization
[09:03:21][V][json:056]: Size after shrink 88 bytes
[09:03:21][V][component:204]: Component light took a long time for an operation (0.05 s).
[09:03:21][V][component:205]: Components should block for at most 20-30ms.
[09:03:21][D][ratgdo:151]: [104598] received mine: rolling=0000657 fixed=0200000539 data=00000081


WARNING smallgarageratgdo.local: Connection error occurred: [Errno 54] Connection reset by peer
INFO Processing unexpected disconnect from ESPHome API for smallgarageratgdo.local
WARNING Disconnected from API
INFO Successfully connected to smallgarageratgdo.local

```

**Comment by bdraco** [comment]
Lock remotes does not appear to have the same issue, only flipping the light seems to cause it

**Comment by bdraco** [comment]
going to remove 
```
logger:
  level: VERBOSE

```
from my test device and downgrade ESP_LOGD to ESP_LOGV and see if I can still replicate the crash
https://github.com/ESPHome-RATGDO/esphome-ratgdo/pull/17

**Comment by bdraco** [comment]
```
[09:17:32][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:32][D][light:036]: 'Light' Setting:
[09:17:32][D][light:047]:   State: OFF
[09:17:32][D][number:012]: 'Rolling code counter': Sending state 2322.000000
[09:17:32][D][light:036]: 'Light' Setting:
[09:17:32][D][light:047]:   State: ON
[09:17:32][D][number:012]: 'Rolling code counter': Sending state 2323.000000
[09:17:32][D][light:036]: 'Light' Setting:
[09:17:32][D][light:047]:   State: OFF
[09:17:32][D][number:012]: 'Rolling code counter': Sending state 2324.000000
[09:17:32][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:32][D][light:036]: 'Light' Setting:
[09:17:32][D][light:047]:   State: ON
[09:17:32][D][number:012]: 'Rolling code counter': Sending state 2325.000000
[09:17:32][D][light:036]: 'Light' Setting:
[09:17:32][D][light:047]:   State: OFF
[09:17:32][D][number:012]: 'Rolling code counter': Sending state 2326.000000
[09:17:32][D][light:036]: 'Light' Setting:
[09:17:32][D][light:047]:   State: ON
[09:17:32][D][number:012]: 'Rolling code counter': Sending state 2327.000000
[09:17:32][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:32][D][light:036]: 'Light' Setting:
[09:17:32][D][light:047]:   State: OFF
[09:17:32][D][number:012]: 'Rolling code counter': Sending state 2328.000000
[09:17:32][D][light:036]: 'Light' Setting:
[09:17:32][D][light:047]:   State: ON
[09:17:32][D][number:012]: 'Rolling code counter': Sending state 2329.000000
[09:17:35][D][light:036]: 'Light' Setting:
[09:17:35][D][light:047]:   State: OFF
[09:17:35][D][number:012]: 'Rolling code counter': Sending state 2330.000000
[09:17:35][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:35][D][light:036]: 'Light' Setting:
[09:17:35][D][light:047]:   State: ON
[09:17:35][D][number:012]: 'Rolling code counter': Sending state 2331.000000
[09:17:35][D][light:036]: 'Light' Setting:
[09:17:35][D][light:047]:   State: OFF
[09:17:35][D][number:012]: 'Rolling code counter': Sending state 2332.000000
[09:17:35][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: ON
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2333.000000
[09:17:36][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: OFF
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2334.000000
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: ON
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2335.000000
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: OFF
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2336.000000
[09:17:36][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: ON
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2337.000000
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: OFF
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2338.000000
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: ON
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2339.000000
[09:17:36][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: OFF
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2340.000000
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: ON
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2341.000000
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: OFF
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2342.000000
[09:17:36][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:36][D][light:036]: 'Light' Setting:
[09:17:36][D][light:047]:   State: ON
[09:17:36][D][number:012]: 'Rolling code counter': Sending state 2343.000000
[09:17:37][D][light:036]: 'Light' Setting:
[09:17:37][D][light:047]:   State: OFF
[09:17:37][D][number:012]: 'Rolling code counter': Sending state 2344.000000
[09:17:37][D][light:036]: 'Light' Setting:
[09:17:37][D][light:047]:   State: ON
[09:17:37][D][number:012]: 'Rolling code counter': Sending state 2345.000000
[09:17:38][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: OFF
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2346.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: ON
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2347.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: OFF
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2348.000000
[09:17:38][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: ON
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2349.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: OFF
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2350.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: ON
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2351.000000
[09:17:38][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: OFF
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2352.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: ON
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2353.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: OFF
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2354.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2355.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: ON
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2356.000000
[09:17:38][D][light:036]: 'Light' Setting:
[09:17:38][D][light:047]:   State: OFF
[09:17:38][D][number:012]: 'Rolling code counter': Sending state 2357.000000
[09:17:38][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:39][D][light:036]: 'Light' Setting:
[09:17:39][D][light:047]:   State: ON
[09:17:39][D][number:012]: 'Rolling code counter': Sending state 2358.000000
[09:17:39][D][light:036]: 'Light' Setting:
[09:17:39][D][light:047]:   State: OFF
[09:17:39][D][number:012]: 'Rolling code counter': Sending state 2359.000000
[09:17:39][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:39][D][light:036]: 'Light' Setting:
[09:17:39][D][light:047]:   State: ON
[09:17:39][D][number:012]: 'Rolling code counter': Sending state 2360.000000
[09:17:39][D][light:036]: 'Light' Setting:
[09:17:39][D][light:047]:   State: OFF
[09:17:39][D][number:012]: 'Rolling code counter': Sending state 2361.000000
[09:17:39][D][light:036]: 'Light' Setting:
[09:17:39][D][light:047]:   State: ON
[09:17:39][D][number:012]: 'Rolling code counter': Sending state 2362.000000
[09:17:39][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:39][D][light:036]: 'Light' Setting:
[09:17:39][D][light:047]:   State: OFF
[09:17:39][D][number:012]: 'Rolling code counter': Sending state 2363.000000
[09:17:39][D][light:036]: 'Light' Setting:
[09:17:39][D][light:047]:   State: ON
[09:17:39][D][number:012]: 'Rolling code counter': Sending state 2364.000000
[09:17:39][D][light:036]: 'Light' Setting:
[09:17:39][D][light:047]:   State: OFF
[09:17:39][D][number:012]: 'Rolling code counter': Sending state 2365.000000
[09:17:40][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:40][D][light:036]: 'Light' Setting:
[09:17:40][D][light:047]:   State: ON
[09:17:40][D][number:012]: 'Rolling code counter': Sending state 2366.000000
[09:17:40][D][light:036]: 'Light' Setting:
[09:17:40][D][light:047]:   State: OFF
[09:17:40][D][number:012]: 'Rolling code counter': Sending state 2367.000000
[09:17:40][D][light:036]: 'Light' Setting:
[09:17:40][D][light:047]:   State: ON
[09:17:40][D][number:012]: 'Rolling code counter': Sending state 2368.000000
[09:17:40][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:40][D][light:036]: 'Light' Setting:
[09:17:40][D][light:047]:   State: OFF
[09:17:40][D][number:012]: 'Rolling code counter': Sending state 2369.000000
[09:17:40][D][light:036]: 'Light' Setting:
[09:17:40][D][light:047]:   State: ON
[09:17:40][D][number:012]: 'Rolling code counter': Sending state 2370.000000
[09:17:40][D][light:036]: 'Light' Setting:
[09:17:40][D][light:047]:   State: OFF
[09:17:40][D][number:012]: 'Rolling code counter': Sending state 2371.000000
[09:17:40][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:40][D][light:036]: 'Light' Setting:
[09:17:40][D][light:047]:   State: ON
[09:17:40][D][number:012]: 'Rolling code counter': Sending state 2372.000000
[09:17:41][D][light:036]: 'Light' Setting:
[09:17:41][D][light:047]:   State: OFF
[09:17:41][D][number:012]: 'Rolling code counter': Sending state 2373.000000
[09:17:41][D][light:036]: 'Light' Setting:
[09:17:41][D][light:047]:   State: ON
[09:17:41][D][number:012]: 'Rolling code counter': Sending state 2374.000000
[09:17:41][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:41][D][light:036]: 'Light' Setting:
[09:17:41][D][light:047]:   State: OFF
[09:17:41][D][number:012]: 'Rolling code counter': Sending state 2375.000000
[09:17:41][D][light:036]: 'Light' Setting:
[09:17:41][D][light:047]:   State: ON
[09:17:41][D][number:012]: 'Rolling code counter': Sending state 2376.000000
[09:17:41][D][light:036]: 'Light' Setting:
[09:17:41][D][light:047]:   State: OFF
[09:17:41][D][number:012]: 'Rolling code counter': Sending state 2377.000000
[09:17:41][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:41][D][light:036]: 'Light' Setting:
[09:17:41][D][light:047]:   State: ON
[09:17:41][D][number:012]: 'Rolling code counter': Sending state 2378.000000
[09:17:41][D][light:036]: 'Light' Setting:
[09:17:41][D][light:047]:   State: OFF
[09:17:41][D][number:012]: 'Rolling code counter': Sending state 2379.000000
[09:17:41][D][light:036]: 'Light' Setting:
[09:17:41][D][light:047]:   State: ON
[09:17:42][D][number:012]: 'Rolling code counter': Sending state 2380.000000
[09:17:42][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:42][D][light:036]: 'Light' Setting:
[09:17:42][D][light:047]:   State: OFF
[09:17:42][D][number:012]: 'Rolling code counter': Sending state 2381.000000
[09:17:42][D][light:036]: 'Light' Setting:
[09:17:42][D][light:047]:   State: ON
[09:17:42][D][number:012]: 'Rolling code counter': Sending state 2382.000000
[09:17:42][D][light:036]: 'Light' Setting:
[09:17:42][D][light:047]:   State: OFF
[09:17:42][D][number:012]: 'Rolling code counter': Sending state 2383.000000
[09:17:42][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:42][D][light:036]: 'Light' Setting:
[09:17:42][D][light:047]:   State: ON
[09:17:42][D][number:012]: 'Rolling code counter': Sending state 2384.000000
[09:17:42][D][light:036]: 'Light' Setting:
[09:17:42][D][light:047]:   State: OFF
[09:17:42][D][number:012]: 'Rolling code counter': Sending state 2385.000000
[09:17:42][D][light:036]: 'Light' Setting:
[09:17:42][D][light:047]:   State: ON
[09:17:42][D][number:012]: 'Rolling code counter': Sending state 2386.000000
[09:17:43][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:43][D][light:036]: 'Light' Setting:
[09:17:43][D][light:047]:   State: OFF
[09:17:43][D][number:012]: 'Rolling code counter': Sending state 2387.000000
[09:17:43][D][light:036]: 'Light' Setting:
[09:17:43][D][light:047]:   State: ON
[09:17:43][D][number:012]: 'Rolling code counter': Sending state 2388.000000
[09:17:43][D][light:036]: 'Light' Setting:
[09:17:43][D][light:047]:   State: OFF
[09:17:43][D][number:012]: 'Rolling code counter': Sending state 2389.000000
[09:17:43][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:43][D][light:036]: 'Light' Setting:
[09:17:43][D][light:047]:   State: ON
[09:17:43][D][number:012]: 'Rolling code counter': Sending state 2390.000000
[09:17:43][D][light:036]: 'Light' Setting:
[09:17:43][D][light:047]:   State: OFF
[09:17:43][D][number:012]: 'Rolling code counter': Sending state 2391.000000
[09:17:43][D][light:036]: 'Light' Setting:
[09:17:43][D][light:047]:   State: ON
[09:17:43][D][number:012]: 'Rolling code counter': Sending state 2392.000000
[09:17:43][D][esp8266.preferences:238]: Saving preferences to flash...
[09:17:43][D][light:036]: 'Light' Setting:
[09:17:43][D][light:047]:   State: OFF
[09:17:44][D][number:012]: 'Rolling code counter': Sending state 2393.000000
[09:17:44][D][light:036]: 'Light' Setting:
[09:17:44][D][light:047]:   State: ON
[09:17:44][D][number:012]: 'Rolling code counter': Sending state 2394.000000
[09:17:44][D][light:036]: 'Light' Setting:
[09:17:44][D][light:047]:   State: OFF
[09:17:44][D][number:012]: 'Rolling code counter': Sending state 2395.000000
[09:17:44][D][esp8266.preferences:238]: Saving preferences to flash...


```

Can still replicate the issue.

**Comment by bdraco** [comment]
Maybe since the light resyncs state plus the defer for the write overloads the 8266. Maybe we need a debounce for that

**Comment by bdraco** [comment]
Will try flashing a local chip and seeing if I can replicate it. If not will get the ladder out and hook up a cable to get a console trace

**Comment by bdraco** [comment]
Doesn't happen on a test chip with no GDO connected.

Going to try to narrow it down before I get out the ladder

**Comment by bdraco** [comment]
commenting out the deferred flash write fixes the crash

**Comment by bdraco** [comment]
https://github.com/ESPHome-RATGDO/esphome-ratgdo/pull/18 should fix the crash but it introduces a very tiny chance of being more out of sync on rebooting

**Review by bdraco — APPROVED** [review]
Tested 👍

**Comment by mariusmuja** [comment]
Great job on figuring this out! I tried to replicate on my GDO, but was unable to... not too surprising since different 8622s can have different flash chips with different timings. Also, I'm not using Paul's circuit (I tried to order one but I couldn't get Square to accept my Canadian shipping address, so I made my own).

A possible solution would be to defer the flash write with a named timeout, such that another flash sync that gets scheduled within that timeout would cancel out the pending one first.

essentially replace:
```cpp
defer([=] {
      this->lastSyncedRollingCodeCounter = this->rollingCodeCounter;
      global_preferences->sync();
});
```
with:
```cpp
set_timeout("sync_preferences", 500, [=] {
        this->lastSyncedRollingCodeCounter = this->rollingCodeCounter;
        global_preferences->sync();
});
```

increase/decrease the 500ms timeout until no longer crashing, but it's hard to know what would happen on other chips...

Anyway, I'm happy with the solution in #18.

**Comment by bdraco** [comment]
I'm happy with the solution in #18. I couldn't get it to fail with manual power pulls. I think the failure risk is very very low

**Pull request merged**